### ISSUE TITLE
Twemproxy hangs parsing nested Array responses from Redis

### ISSUE THREAD

**Issue by jhwillett** [twitter/twemproxy#537]
Twemproxy hangs when processing a Redis Lua script via EVALwhich returns nested sub-arrays. 
This is easy to reproduce.

Setup:
* On Mac OS 10.12.6.
* Running redis-server 2.8.24 at localhost:7379 (installed via Homebrew).
* Running nutcracker-0.4.1 at localhost:221221 which is configured with:
```
development:
  redis:            true
  auto_eject_hosts: false
  listen:           127.0.0.1:22121
  servers:
    - 127.0.0.1:7379:1
```

Demo of direct connection to Redis being happy:
```
$ redis-cli -p 7379
127.0.0.1:7379> EVAL 'return {1,2,3}' 1 x
1) (integer) 1
2) (integer) 2
3) (integer) 3
127.0.0.1:7379> EVAL 'return {1,{2},3}' 1 x
1) (integer) 1
2) 1) (integer) 2
3) (integer) 3
127.0.0.1:7379>
```

Demo of connection through Twemproxy being unhappy:
```
$ redis-cli -p 221221
127.0.0.1:221221> EVAL 'return {1,2,3}' 1 x
1) (integer) 1
2) (integer) 2
3) (integer) 3
127.0.0.1:221221> EVAL 'return {1,{2},3}' 1 x
^C
```
The last command hangs as long as I was willing to wait (several minutes), until I cancel it.

I see no messages in the twemproxy logs during this time.

I notice that src/proto/nc_redis.c has special cases for nested multi bulk reply element from sscan/hscan/zscan.  Perhaps eval and evalsha are simply missing the special case treatment?

I apologize for phrasing this as a bug report instead of as a pull request with a fix, but I am afraid it may be a long time before I can dig deeper.

Thank you for a wonderful tool.  It is a critical piece of our infrastructure at ProsperWorks and has made our lives much easier.

**Comment by TysonAndre** [comment]
I think the code in question is https://github.com/twitter/twemproxy/blob/v0.4.1/src/proto/nc_redis.c#L2084-L2107 and possibly others

Changing from a tail recursive parser (with goto) to a real recursive parser for multi-bulk calls is possibly how that would be handled - I assume the performance of sscan/hscan/zscan isn't important to the vast majority of applications, and recursive should still be fairly efficient

I think I had the same workaround of JSON encoding return values (or using some other portable encoding) in lua scripts

**Comment by TysonAndre** [comment]
Fixed in #612 - it turns out it didn't make sense to make it recursive - the only state that needed to be tracked was

1. Whether this was currently at the top level
2. The number of remaining elements to parse, increasing when new arrays are encountered

This will be part of an upcoming 0.5.0 release
